### ISSUE TITLE
Flags definition policy

### ISSUE THREAD

**Issue by kinow** [BSC-ES/autosubmit#1383]
In GitLab by @naiaraAM on Aug 5, 2024, 09:57

As Bruno realised we kind of have a mess on flags definition for commands. The questions is, should we leave the single-dash-double-letter (as ``-ft`` for ``--fiter_type``) with expanded double-dash option, single-letter with with double-dash (as ``-b`` for ``--git_branch``) or just double-dash (as ``--hide_groups``)?

@dbeltrankyl @edgano @mcastril

**Comment by kinow** [comment]
marked this issue as related to #1237

**Comment by kinow** [comment]
Thanks for creating the issue, @naiaraAM !

I linked @ainagaya 's issue about `--all` vs. `-all`.

The convention on Linux is `-` for single/short ops, and `--` for long options.

Some commands accept you to combine short options into a single one (e.g. `tar -zxvf` is the same as `tar -z -x -v -f`, `docker -ti` is equivalent to `docker -t -i`, etc.).

For me the key is consistency, and avoiding surprises to user. For example, if most of our commands have `-s` as a short option for `--save`, we should avoid `-s` in other commands unless that option is really obvious.

And when we add a new option that might in the future have a variant (for example, if we add `--save-something` and it is likely we will have `--save-something-else`), then in cases like this we must avoid the short options.

There is nothing wrong with short options. Some think users will get annoyed by not having a short option, but in reality they get used to. `docker` has the `--build-args`, and no `-b` (as there are `--builder`, `--build-arg`, `--build-context` -- i.e. how would you know which `b` it is? And adding `-ba` would break if there was ever another `--build-a-something` option). Users get used to this, and might appreciate the consistency.

`git` has the same, `git commit --squash`, `--short`. The `-s` is the `signoff`, which is consistent as both `git commit` and `git tag` have `-s` to sign the commits.

Some references for `-` short, and `--` long.

- From GNU, “25.1.1 Program Argument Syntax Conventions” https://www.gnu.org/software/libc/manual/html_node/Argument-Syntax.html
  - Quoting “Arguments are options if they begin with a hyphen delimiter (‘-’).”
  - “Multiple options may follow a hyphen delimiter in a single token if the options do not take arguments. Thus, ‘-abc’ is equivalent to ‘-a -b -c’.”
  - “Option names are single alphanumeric characters (as for isalnum; see Classification of Characters).”
- From IEEE/POSIX “12. Utility Conventions” https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap12.html
  - Quoting, “Each option name should be a single alphanumeric character (the alnum character classification) from the portable character set.”
- The Open Source “Command Line Interface Guidelines” https://clig.dev/
  - They have a section for “Consistency across programs” https://clig.dev/#consistency-across-programs
  - And more, e.g.

![image](https://earth.bsc.es/gitlab/es/autosubmit/uploads/1b4e2916fb685d52341899b12e13eaec/image.png)

So we are not following POSIX, not following GNU, nor following what other tools like git/docker/cylc/cwltool/etc do, nor recent community/open source guidelines like clig.dev.

IMHO, adding a long option is an easy task, we just give it a simple and descriptive name. For short options, again, IMHO, we should discuss maybe using it for the most used options, in a consistent way, and not be afraid to force users to use the long version when we have a good reason for not using a single option arg.

**Comment by kinow** [comment]
In GitLab by @edgano on Aug 5, 2024, 10:28

+1 with @kinow 
and I would like to highlight this:
```
IMHO, adding a long option is an easy task, we just give it a simple and descriptive name. For short options,
 again, IMHO, we should discuss maybe using it for the most used options, in a consistent way, and not be 
afraid to force users to use the long version when we have a good reason for not using a single option arg.
```

And again, I think we should no be afraid, just be consistent and have a clear idea/view of what we have and what we want. (and document all :stuck_out_tongue_closed_eyes: )

**Comment by kinow** [comment]
Yes, adding is easy, but to remove, I think we would have to add a deprecation warning, warn users, and wait until a long time down the 4.x release line, or be more pedantic and wait for autosubmit 5.x :sweat:

**Comment by kinow** [comment]
In GitLab by @mcastril on Aug 9, 2024, 17:59

I completely agree, and in general, I am more fond of long arguments, unless the short one is quite obvious.

One example is the rsync command used in many of our workflows. I have always suggested using `rsync --dry-run --perms` rather than `rsync -n -p` because the short flag is not always obvious, and it's more understandable to anyone who is looking at the code.

I must admit that I have never had issues against `-ft` or `-np` in Autosubmit because these are quite clear and it's Autosubmit language. The issue is when we break our own rules. For example, `-all` would clearly be a mistake.

**Comment by kinow** [comment]
I also prefer to use the long arguments when writing scripts, documentation, or things others may work on :slight_smile: And I never memorized the `rsync` arguments, so anytime I see a short argument, I need to open `rsync`'s man page to translate that... so thanks for doing that! :pray: 

>I must admit that I have never had issues against `-ft` or `-np` in Autosubmit because these are quite clear and it's Autosubmit language. The issue is when we break our own rules. For example, `-all` would clearly be a mistake.

I agree these are clear within Autosubmit, but I think the issue is not only breaking our own rules.

In the long term, we may run out of options, as we continue to grow and get more options/arguments, eventually there will be conflicts with `-ft` or `-np`, and we might run in situations where we need to use `-ft2` or give it some other abbreviation.

There is also the point that it could be better for users if Autosubmit followed open standards/conventions on argument parsing, instead of having its own language (especially as we would have to be very disciplined to keep this language well concise).

I believe following what is done in other famous commands (and in many workflow managers) would simplify for developers, and at the same time help future external (non-BSC) users of Autosubmit, IMO.

**Comment by kinow** [comment]
In GitLab by @edgano on Aug 12, 2024, 14:40

I agree to sit down and discuss the guidelines for the flags (my point is to follow standards or strategies used by other software/workflow managers)

I can see that in the future we want to modify multiple params from the command line. It give the user more flexibility and it makes the executions faster.

as a little comment (not the full strategy), what we define in the command line (`--input`) in Nf, we will be able to access under the `params.input`

And I can see it in the case we want to modify params of the config files (yaml) ->  `params.default.chunksize` (for example)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Oct 3, 2024, 08:45

changed due date to October 10, 2024

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Oct 3, 2024, 08:45

removed due date

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Oct 3, 2024, 10:17

changed due date to October 24, 2024

**Comment by pablogoitia** [comment]
@pablogoitia reminder for checking this
